integrations/tensorflow/iree_tf_compiler/TF/SavedModelToIreeABI.cpp

```diff
@@ -0,0 +1,577 @@
+          json::Array nvRecord;
+          nvRecord.push_back(child.skey);
+          nvRecord.push_back(child.createReflectionType());
+          typeRecord.push_back(json::Value(std::move(nvRecord)));
+        }
+        return json::Value(std::move(typeRecord));
+      }
+      default:
+        llvm_unreachable("Unsupported LevelType");
+    }
+
+    return json::Value(nullptr);
+  }
+
+  // Recursively emits argument loads by processing all children and
+  // populating callArgs with the Values of leaves.
+  void emitDereferenceArgs(Location loc, OpBuilder &builder, Value thisValue,
+                           SmallVector<Value> &callArgs) {
+    // Terminal.
+    if (type == LevelType::Value) {
+      assert(valueIndex < callArgs.size() && "mismatched number of call args");
+      assert(!callArgs[valueIndex] && "duplicate argument bindings");
+      callArgs[valueIndex] = thisValue;
+      return;
+    }
+
+    // Recurse into sequence (index can be sparse on child ikey).
+    if (type == LevelType::List || type == LevelType::Tuple) {
+      for (StructureLevel &child : children) {
+        Value childValue =
+            child.emitGetFromList(loc, builder, thisValue, child.ikey);
+        child.emitDereferenceArgs(loc, builder, childValue, callArgs);
+      }
+      return;
+    }
+
+    // Recurse into dict (modeled as a dense tuple of children).
+    if (type == LevelType::Dict) {
+      for (auto it : llvm::enumerate(children)) {
+        StructureLevel &child = it.value();
+        Value childValue =
+            child.emitGetFromList(loc, builder, thisValue, it.index());
+        child.emitDereferenceArgs(loc, builder, childValue, callArgs);
+      }
+      return;
+    }
+    llvm_unreachable("unhandled StructureLevel type");
+  }
+
+  // Emits operations to recursively create this structure from the given
+  // ValueRange of flattened values.
+  Value emitCreateReturns(Location loc, OpBuilder &builder,
+                          ResultRange &callReturns) {
+    // Terminal.
+    if (type == LevelType::Value) {
+      assert(valueIndex < callReturns.size() &&
+             "mismatched number of call returns");
+      return callReturns[valueIndex];
+    }
+    // Recurse into sequence (index can be sparse on child ikey).
+    if (type == LevelType::List || type == LevelType::Tuple) {
+      Value listSizeValue =
+          builder.create<ConstantOp>(loc, builder.getIndexType(),
+                                     builder.getIndexAttr(getNeededListSize()));
+      Value listValue = builder.create<iree_compiler::IREE::ListCreateOp>(
+          loc, getIrType(builder), listSizeValue);
+      for (StructureLevel &child : children) {
+        Value childValue = child.emitCreateReturns(loc, builder, callReturns);
+        Value indexValue = builder.create<ConstantOp>(
+            loc, builder.getIndexType(), builder.getIndexAttr(child.ikey));
+        // TODO: ListSetOp does not seem to have side effects.
```

Comment:
it is not marked pure and should be conservatively assumed to have side-effects - do you have any more info on this?